### PR TITLE
SCIM FE

### LINKED ISSUE #43132: [FE] UI for enabling SCIM and displaying the API key
[figma](https://www.figma.com/design/E42ezp6P3kY4PJj0r5tIOM/Support-SCIM-for-user-provisioning-and-deprovisioning?node-id=2095-1152)

Branch: `nm-scim-be`

The relevant bits are:
- `scim-enabled` and `scim-base-url` settings
- `GET /api/ee/scim/api_key` for fetching an existing (masked) SCIM API key
- `POST /api/ee/scim/api_key` for creating or recreating the key

### PR DESCRIPTION
**Pull request by sloansparger**
Closes #43132

### Description

Implementation for the FE for SCIM following the designs shown here: [figma](https://www.figma.com/design/E42ezp6P3kY4PJj0r5tIOM/Support-SCIM-for-user-provisioning-and-deprovisioning?node-id=2095-1152)

### How to verify

- go to admin settings -> authentication
-  click on the new "user provisioning" tab
- enable scim for the first time
- try disabling scim + enabling again
- try regenerating a token
- try enabling SAML SSO to engage different parts of the experience
- test out on both OSS and EE

### Demo

TODO

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: src/metabase/public_settings.clj [hunk 1]
```diff
@@ -741,6 +741,7 @@ See [fonts](../configuring-metabase/fonts.md).")
                       :hosting                        (premium-features/is-hosted?)
                       :official_collections           (premium-features/enable-official-collections?)
                       :sandboxes                      (premium-features/enable-sandboxes?)
+                      :scim                           (premium-features/enable-scim?)
```
Comment (sloansparger) [inline src/metabase/public_settings.clj:744]: \\(★ω★)/
i am a clojure god

File: enterprise/frontend/src/metabase-enterprise/user_provisioning/components/ScimInputs.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  label: string;
+  value: string;
+  disabled?: boolean;
+}) => (
+  <CopyTextInput
+    label={label}
+    value={value}
+    readOnly
+    disabled={disabled}
+    styles={getTextInputStyles({ masked: false, disabled })}
```
[the remark this answers is not in the document]
Comment (sloansparger) [reply enterprise/frontend/src/metabase-enterprise/user_provisioning/components/ScimInputs.tsx:39]: Let's sync on this on monday. I'm a bit lost on what the best practice is currently.

File: enterprise/frontend/src/metabase-enterprise/api/api.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { Api } from "metabase/api";
+
+import { ENTERPRISE_TAG_TYPES } from "./tags";
+
+export const EnterpriseApi = Api.enhanceEndpoints({
+  addTagTypes: ENTERPRISE_TAG_TYPES,
```
Comment (npfitz) [inline enterprise/frontend/src/metabase-enterprise/api/api.ts:6]: 🔥

File: enterprise/frontend/src/metabase-enterprise/user_provisioning/components/UserProvisioning.tsx [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+  const samlUserProvisioningEnabled = useSetting(
+    "saml-user-provisioning-enabled?" as any,
+  );
+  const showSamlWarning = samlUserProvisioningEnabled && !isScimInitialized;
+
+  const handleScimEnabledChange = async (enabled: boolean) => {
+    const setting = fields["scim-enabled"];
+    // HACK: this is the only way to conditionally update the UI
+    // in response to the setting being successfully set
+    setting.onChanged = async () => {
```
Comment (npfitz) [inline enterprise/frontend/src/metabase-enterprise/user_provisioning/components/UserProvisioning.tsx:107]: Lets rubber duck on this. It works, but mutating the setting feels a little gross

File: enterprise/frontend/src/metabase-enterprise/user_provisioning/components/UserProvisioningModals.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import { ScimTextWarning } from "./ScimTextWarning";
+
+type BaseUserProvisiongModalProps = Pick<ModalProps, "opened" | "onClose">;
+
+interface ScimModalProps extends BaseUserProvisiongModalProps {
+  title: string;
+  children: React.ReactNode;
+}
+
+const ScimModal = ({ opened, onClose, title, children }: ScimModalProps) => (
```
Comment (npfitz) [inline enterprise/frontend/src/metabase-enterprise/user_provisioning/components/UserProvisioningModals.tsx:18]: I hate that we need to do this. But I get it 🤦‍♂️

**Review by npfitz — APPROVED** [review]
Really only 1 comment on mutating the setting, but I wouldn't consider it blocking. I do find the OSS API Keys experience a bit sub par since it's not consistent with the rest of the pages though
EmMRvrRHiF

**Comment by sloansparger** [comment]
> Really only 1 comment on mutating the setting, but I wouldn't consider it blocking. I do find the OSS API Keys experience a bit sub par since it's not consistent with the rest of the pages though EmMRvrRHiF [ EmMRvrRHiF ](https://private-user-images.githubusercontent.com/1328979/349235450-e309aa60-23c0-46cb-9b62-c2f8907c623d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExNzA0NzksIm5iZiI6MTcyMTE3MDE3OSwicGF0aCI6Ii8xMzI4OTc5LzM0OTIzNTQ1MC1lMzA5YWE2MC0yM2MwLTQ2Y2ItOWI2Mi1jMmY4OTA3YzYyM2QuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTZUMjI0OTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTVlZTIyZTQ5MTljODUyNmRkNjg1ZWRmZjlhNjM4Y2ExODM2MjY2YjdlODhjZGQxNDQ2Njc4ZTYxMzkzMTMxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.iYSFHRdOQgLIQV4tVEn7BeB6XowLpd8wPfLR--5-Be4) [ ](https://private-user-images.githubusercontent.com/1328979/349235450-e309aa60-23c0-46cb-9b62-c2f8907c623d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExNzA0NzksIm5iZiI6MTcyMTE3MDE3OSwicGF0aCI6Ii8xMzI4OTc5LzM0OTIzNTQ1MC1lMzA5YWE2MC0yM2MwLTQ2Y2ItOWI2Mi1jMmY4OTA3YzYyM2QuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTZUMjI0OTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTVlZTIyZTQ5MTljODUyNmRkNjg1ZWRmZjlhNjM4Y2ExODM2MjY2YjdlODhjZGQxNDQ2Njc4ZTYxMzkzMTMxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.iYSFHRdOQgLIQV4tVEn7BeB6XowLpd8wPfLR--5-Be4)

All fixed! Good catch!

File: frontend/src/metabase/admin/settings/app/components/SettingsEditor/SettingsEditor.jsx [hunk 6]
```diff
@@ -92,27 +99,29 @@ class SettingsEditor extends Component {
...
-          oldValue,
-          newValue,
-          settingValues,
-          this.handleChangeSetting,
-        );
+        await setting.onChanged(...handlerParams);
+      }
+
+      if (options?.onChanged) {
+        await options.onChanged(...handlerParams);
```
Comment (npfitz) [inline frontend/src/metabase/admin/settings/app/components/SettingsEditor/SettingsEditor.jsx:124]: I like it. I also like that it doesn't overwrite the `onChanged` in the setting itself 👍

**Review by npfitz — APPROVED** [review]

**Pull request merged**